File: src/main/java/arcana/utils/Pair.java
package arcana.utils;

import java.util.Objects;

public final class Pair<A, B>{
	public A a;
	public B b;
	
	public Pair(A first, B second){
		this.a = first;
		this.b = second;
	}
	
	public static <A, B> Pair<A, B> of(A first, B second){
		return new Pair<>(first, second);
	}

	public Pair<B, A> flip(){
		return of(b, a);
	}
	
	public boolean contains(Object obj){
		return a == obj || b == obj;
	}
	
	public boolean equals(Object o){
		if(this == o)
			return true;
		if(!(o instanceof Pair))
			return false;
		Pair<?, ?> pair = (Pair<?, ?>)o;
		return Objects.equals(a, pair.a) && Objects.equals(b, pair.b);
	}
	
	public int hashCode(){
		return Objects.hash(a, b);
	}
}

File: src/main/kotlin/arcana/utils/Pair.kt
package arcana.utils

import java.util.*

class Pair<A, B>(var a: A, var b: B) {
    fun flip(): Pair<B, A> {
        return of(b, a)
    }

    operator fun contains(obj: Any): Boolean {
        return a === obj || b === obj
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Pair<*, *>) return false
        val pair = other
        return a == pair.a && b == pair.b
    }

    override fun hashCode(): Int {
        return Objects.hash(a, b)
    }

    companion object {
        fun <A, B> of(first: A, second: B): Pair<A, B> {
            return Pair(first, second)
        }
    }
}